experimental: flux job command, job-manager module

garlick opened the pull request:
This is a toy "front end" for the new exec system.  Don't get your hopes up!

There are three parts here:
* `flux-manager` module for job ingest
* `flux job run` client for submitting jobs
* `libjob` - a placeholder for a new API for job submission/control

The API at this point only includes
```c
enum {
    FLUX_JOB_NOVERIFY_SIGNATURE = 1,    // don't verify signer == requestor
};

typedef uint64_t flux_jobid_t;

flux_future_t *flux_job_run (flux_t *h, const char *J, int flags);
int flux_job_run_get_id (flux_future_t *f, flux_jobid_t *id);
```

`J` is a signed payload, expected to be jobspec, although for now only the signature is verified; the payload is not parsed, so you could stuff anything in there.

The command line allows jobspec (files or stdin) to be signed and submitted, and prints jobid(s) on stdout.
```
Usage: flux-job run [OPTIONS] jobspec
Run job(s)
  -U, --no-verify-signature
                         job-module should defer signature verification
  -c, --security-config=pattern
                         Use non-default security config glob
  -f, --fanout=N         Run at most N RPCs in parallel
  -h, --help             Display this message.
  -r, --repeat=N         Run N instances of jobspec
  -w, --words            Print jobid as a six word phrase
```

The `job-manager` module accepts `flux_job_run()` requests, assigns jobids (FLUIDs), and creates a KVS hierarchy under `job` that uses the FLUID "dothex" encoding to spread jobids over the KVS keyspace.   Thus far only the `J-signed` key specified in [RFC 16](https://github.com/flux-framework/rfc/blob/master/spec_16.adoc) is populated.

I worked on performance a bit.  In the `job-manager` module, KVS updates are batched for up to 10ms, so a burst of job requests may be committed together.  In the client, the code tries to keep `--fanout` RPCs in flight simultaneously.   For a test case of
```
flux job run  --repeat 100000 --no-verify-signature ./myjob
```
where "myjob" is example 2.1 from [RFC 14](https://github.com/flux-framework/rfc/blob/master/spec_14.adoc), and default fanout of 256, the ingest rate is about 10K jobs per second.  With signature verification (configured for munge), about 3K jobs per second.   The KVS contains lots of entries like this:
```
job.0000.0067.3200.0005.J-signed = dmVyc2lvbgBpMQBtZWNoYW5pc20Ac211bmdlAHVzZX...
job.0000.0067.3200.0006.J-signed = dmVyc2lvbgBpMQBtZWNoYW5pc20Ac211bmdlAHVzZX...
job.0000.0067.3200.0007.J-signed = dmVyc2lvbgBpMQBtZWNoYW5pc20Ac211bmdlAHVzZX...
job.0000.0067.3200.0008.J-signed = dmVyc2lvbgBpMQBtZWNoYW5pc20Ac211bmdlAHVzZX...
job.0000.0067.3300.0000.J-signed = dmVyc2lvbgBpMQBtZWNoYW5pc20Ac211bmdlAHVzZX...
job.0000.0067.3300.0001.J-signed = dmVyc2lvbgBpMQBtZWNoYW5pc20Ac211bmdlAHVzZX...

```
(the signed blob looks the same because I signed it once and submitted it `--repeat` times).

TODO:
* Discussion(!)
* Add option to individually sign repeat requests so we can see performance impact
* Figure out how to call C++ jobpsec parser from `job-manager` and then see how validating goes.

There was no attempt to write tests or document at this stage.  Maybe we'll scrap the whole thing after we have a talk about it, so I didn't want to put too much into this!

grondo commented:
One quick comment, it might be wise at this point to save the `flux job run` command for later when we've designed the UI for interactive job running (`flux job run`) and job submission (`flux job submit`). This way we can continue to modify the test utility without worrying about perturbing a core user interface with test options, etc.

grondo commented:
Really cool by the way! I really like the delayed kvs flush, and the implementation of that turned out so nice and simple.

In retrospect, it sure would be nice if there was an implementation of `flux_sign_wrap` that returned a future (though we have a circular dependency there sadly). I wonder if the fact that the module blocks when verifying has anything to do with the drastic drop in throughput for that case?

garlick commented:
I hadn't thought about it that this way, but you're right, the munge/unmunge calls encapsulate a synchronous call to the MUNGE daemon, and so will stall our (single threaded) reactor loop each time a job is processed.  It would be interesting to see the performance of the curve signing method in comparison.

Thoughts on how to work around the MUNGE interface?   Some kind of threaded adapter?  A specialized MUNGE client that we could incorporate into our reactor loop? 

Regarding the `flux job run` command, excellent point about needing to carefully design the user interface and not have it grow a bunch of options needed for testing or experimentation!  I'll rename to `flux job testrun` or similar (or we could move it entirely out of the user facing command into `t/job/testrun` or similar).

garlick commented:
Oh just realized the curve method requires a synchronous call to the imp, so I see your point about needing a future or equivalent.

grondo commented:
> I'll rename to flux job testrun or similar (or we could move it entirely out of the user facing command into t/job/testrun or similar).

I'd be fine with `testrun` or `submitbench` under the `flux job` command, could be useful in the future

> Thoughts on how to work around the MUNGE interface? Some kind of threaded adapter? A specialized MUNGE client that we could incorporate into our reactor loop?

Yeah, my only thought was to have a worker pool of `flux_sign_unwrap` threads, perhaps even behind a service which could offer unwrap over an RPC and farm out the work to its threads.

But I like your idea of a specialized MUNGE client much better!

However, even if MUNGE API had an interface that could be hooked into a reactor, the `flux_sign_wrap/unwrap()` interface would need to be replaced with something that could utilize it. Additionally, to avoid circular dependency between libflux-security and libflux-core, the security API can't directly return a `flux_future_t`, but would have to export lower level functions that could then be used to build a single function returning a `flux_future_t` within flux-core... (is that making sense?)

garlick commented:
Makes sense.

Although, heh, look in `t/rpc/util.c` and `t/rpc/rpc.c`.  There is a "test server" implementation in there that starts a callback with a `flux_t` handle that is run on the other end of a  `shmem://`  connector, similar to czmq's zactor.  In that case it allowed the RPC client and server to be tested without a broker.  The callback just registers message handlers and starts the reactor like a comms module.

We could avoid stalling the job-manager thread by implementing something similar that performs the blocking unwrap call in another thread.   Maybe we just need an API in front of that general concept in flux-core?  Then extend to allow a round-robin work crew of individual actors to gain speed?

grondo commented:
> We could avoid stalling the job-manager thread by implementing something similar that performs the blocking unwrap call in another thread. Maybe we just need an API in front of that general concept in flux-core? Then extend to allow a round-robin work crew of individual actors to gain speed?

Yes, that would seem to work as well, though of course it would be nicer if code we control worked non-blocking... But the general concept of shuttling a blocking function to a thread seems like a nice addition to flux-core.

garlick commented:
> Yes, that would seem to work as well, though of course it would be nicer if code we control worked non-blocking... But the general concept of shuttling a blocking function to a thread seems like a nice addition to flux-core.

Agreed, and I was just pleased to remember there was something already semi-worked out, not necessarily arguing that it was the better choice.   Although I find it challenging to work out the right interface to make e.g. `flux_sign_unwrap()` non blocking without unduly complicating flux-security...

grondo commented:
> Although I find it challenging to work out the right interface to make e.g. flux_sign_unwrap() non blocking without unduly complicating flux-security...

Yeah, that is a good point, and of course there are going to be other blocking functions libflux-core API users would like to use...

garlick commented:
Just pushed another (sqashed) set of changes:

* Rename `job-manager` to `job-ingest` per discussion with @grondo.  We'll keep ingest separate from the managment functionality, so that we can load `job-ingest` across an instance and possibly load `job-manager` (which may have a larger memory footprint) on a subset of nodes.
* Issue an event containing a batch of new jobid's when batch KVS commits complete
* Code for committing batches of KVS data reworked - it was a bit too simplistic before and may have responded to users before the KVS commit was complete in some overlapping cases.
* Began to add a sharness test but this is not yet working.

I'll keep going on this next week.

garlick commented:
Changes since the last comment
* Updated to track `flux-security` API changes
* Copious inline comments in `job-ingest` module
* Submitted jobspec is validated, with text error messages returned from parser on invalid jobspec
* Added `flux job id` command to convert between jobid representations
* sharness test works (hopefully in travis, after various issues)

garlick commented:
Just pushed a squashed and cleaned up version of this PR, with more tests and better commit documentation.  Hopefully the coverage will be a little better than it was.

garlick commented:
I think this PR now address the ingest portion of #1332 and is probably a candidate for merging.

One caveat:  this module is not currently reloadable.  For that to work, we will need some way for FLUID generator state to be save/restored (#1545), and I propose we do that in another PR.  That's the only impediment to making this reloadable that I can think of.

grondo commented:
If we do consider merging this now, we should probably wrap up any minor fixes and improvements to flux-core first, merge those and then tag a release.

Once this PR is merged flux-core now has a hard dependency on flux-security and we'd have to package and release that project for use on our systems where flux-core and flux-sched are currently installed.

Alternatively, we could create a "stable" branch off v0.9.0, and fixes destined for "alpha-production" could go on that branch (hopefully not much) as v0.9.1, v0.9.2, and master would become the unstable/experimental branch.

garlick commented:
Oh right, I forgot we don't have a package for flux-security yet.  My vote would be to create a stable branch from 0.9.   However, see "creating stable releases" in [RFC 1](https://github.com/flux-framework/rfc/blob/master/spec_1.adoc).

It mandates one branch per github project, that the main project master is the unstable development branch and, and that github repo clones are the mechanism for creating stable branches.  So if we follow this, we'd clone flux-core to flux-core-0.9 and its master would be the first stable production "branch".

I guess the advantage is a stable repo can have different maintainers, and it's more clear to visitors of your organization site where the stable bits are?

(Ew, that will break our PKG_NAME, PKG_VERSION that is derived from git describe)

grondo commented:
Not only do we not have a flux-security package yet, but installing the package on production systems is not going to be trivial because of the setuid executable. However, I guess a temporary workaround for that issue would be install the IMP without setuid for now until we have something that requires it.

> So if we follow this, we'd clone flux-core to flux-core-0.9 and its master would be the first stable production "branch".

> I guess the advantage is a stable repo can have different maintainers, and it's more clear to visitors of your organization site where the stable bits are?

> (Ew, that will break our PKG_NAME, PKG_VERSION that is derived from git describe)

I guess we could fix the PKG_NAME/VERSION stuff if we go this way, but I'm really on the fence about this approach. Splitting into separate repos does have some good side effects as you mentioned. It also allows for a separate issue tracker for the "stable" vs "development" branch, but do you lose all that context when you retire an old clone? What happens when you have the same issue # in the stable branch vs the development branch and you cherry-pick or merge fixes from stable to master? You basically lose the ability to reference issues in commit comments I think.

I guess I should look over at how this works in practice. 

Thanks for remembering that is in the C4.1 RFC, btw!

garlick commented:
zeromq has

https://github.com/zeromq/libzmq - development
https://github.com/zeromq/zeromq4-1 - 4.1 stable
https://github.com/zeromq/zeromq4-x - 4.0 stable
https://github.com/zeromq/zeromq3-x - 3.0 stable

They all have releases.  Confusingly, libzmq seems to contain 4.2 stable releases.  I guess they don't fork the libzmq repo to zeromqN-x until they have started tagging N+1 releases in libzmq and need another N release?  I hope they don't then remove releases from libzmq as this would break builders.

They all have issues.  The first one I looked at in zeromq4-x contained this comment:
> For these and other issues on 4-x, please raise the issue on zeromq/libzmq,
and if there is a fix already there, you can propose a backport of it to
zeromq4-x (and make a pull request).

I guess you'd have to  use the longhand form to xref issues e.g. flux-framework/flux-core#1543.  I guess that would argue for using longhand in all PR's so they can be applied in multiple repos and still reference the correct issue.

Meh.  Does seem like a pain, but we could try it with `flux-core0-x` and see how much we hate it before we need to do it "for real"...

garlick commented:
I suppose we could also add a `--with-flux-security` configure option here and try to avoid forking...

grondo commented:
> Meh. Does seem like a pain, but we could try it with flux-core0-x and see how much we hate it before we need to do it "for real"...

I guess I could get behind that, though I try to avoid finding out exactly how much I will hate something :wink:

Can we give the repo a reasonable name like `flux-core-0.9` or `flux-core-0.9.x` though? Unless there is some mystical purpose for directly concatenating the name with the version?

grondo commented:
> I suppose we could also add a --with-flux-security configure option here and try to avoid forking...

Ah! The kick-the-can-down-the-road solution! I like it!

garlick commented:
Someone had to suggest it.  When I think of writing more m4 though, it reminds me that pressing the fork button is easier...

garlick commented:
It looks like they wanted to avoid periods in the repo name so `zeromq-4.1` became `zeromq4-1` I guess?  But I just successfully created a `test-4.2` repo so I am ignorant of whatever mysticism is at work.

grondo commented:
> It looks like they wanted to avoid periods in the repo name

I wonder if that went back to CVS naming restrictions??

What is the benefit of forking branches again? I was discussing with @dun and we couldn't come up with anything. (And sorry, should we move this OT discussion to an issue?)

garlick commented:
OK, I'll open an issue on branches vs forked repos and try to list any benefits I am aware of there.

I suppose we just need a way to continue development on 0.9 and that doesn't have to be solved until we have to backport something.

Let's try to get this thread back on track on what needs to happen to get this PR merged.

garlick commented:
Gogle code didn't like periods.  I have no idea if that was a factor.

garlick commented:
I went ahead and added the (opt-in) `configure --with-flux-security` option.

If this option is not provided, then only the instance owner is allowed to submit jobs, and the `J-signed` KVS key is left unpopulated (but otherwise everything works as before).

I tidied up a few other things, added some more inline documentation, and squashed.

garlick commented:
Rebase on current master.
